### PR TITLE
[Rust] Fixed the issue 1816

### LINKED ISSUE #1816: [Rust] `CallingFrame` in wasmedge-sdk
When I use the `CallingFrame` in `wasmedge-sdk`, for example with the method `module_instance()`, the return type is `wasmedge_sys::Instance`, rather than `wasmedge_sdk::Instance`. Such inconsistencies will undoubtedly cause problems and be detrimental to the use of the `wasmedge-sdk`.

@apepkuss Please take a look at this question, thank you very much!

### PR DESCRIPTION
**Pull request by apepkuss**
This PR is to fix #1816. Duo to `cargo nightly clippy` updated to `v0.1.65` in CI, in addition, some new clippy issues are also fixed in this PR.

### CODE REVIEW

**Comment by apepkuss** [comment]
@hydai Could you please help review this PR? Please ignore the failed release workflows, as the dependencies are not published onto `crates.io`. Thanks a lot!

**Comment by hydai** [comment]
Hi @apepkuss 
Just want to check if we do need these failed workflows to be executed in PR. It's highly related to the published version. However, there is no way to use it before publishing it. Is there any solution or practice to make this smooth?

**Comment by apepkuss** [comment]
> Hi @apepkuss Just want to check if we do need these failed workflows to be executed in PR. It's highly related to the published version. However, there is no way to use it before publishing it. Is there any solution or practice to make this smooth?

My idea is to remove the `Dry run` jobs from the release workflows, as the `Publish` jobs have a `dry run` step that can check the code before publishing. What do you think about it?

**Comment by hydai** [comment]
IIRC, we failed the publish because there is no dry-run CI before we tagged.
And we can also find that the versions have dependencies. Maybe we should bump the version per PR? The current rust PR will do too many things inside a single PR. If we can bump the version separately, then it can be tested sequentially. Just like how we do in releasing process.

**Comment by apepkuss** [comment]
> IIRC, we failed the publish because there is no dry-run CI before we tagged. And we can also find that the versions have dependencies. Maybe we should bump the version per PR? The current rust PR will do too many things inside a single PR. If we can bump the version separately, then it can be tested sequentially. Just like how we do in releasing process.

Let me explain more about the version bump and the `dry-run` test:

- The reason for bumping versions in this PR is that we introduced some new objects in `sdk` and `sys` crates; meanwhile, due to the upgrade of `cargo::clippy` used in the CI workflows, the new lint named "result-large-err" requires us to refactor the original design of `wasmedge_types::WasmEdgeResult` that forces the APIs in Rust bindings to support returning new error types. And more importantly, the new result type has already broken the APIs. Therefore, we have to bump the versions in this PR so that we can distinguish between the original design and the new one.

- For the `dry-run` test, you're right, we also run into the same failure caused by the incorrect publishing order. In fact, the real reason for the failure is also the publish operation did not find the correct dependencies on `crates.io`, as we did not publish the dependencies before publishing the high-level crate. Back to the failures in this PR, even if we did not bump the versions, that is, we still used the original versions, the `cargo publish --dry-run` step would also fail, as the dependencies the cargo find out on `crates.io` had no the APIs we need. Therefore, the `dry-run` step will always fail if any API or any design is changed in the low-level dependencies.

**Comment by hydai** [comment]
Okay, so this is the expected behavior of the current workflows.
Just another thing to check, once this PR got merged, do I need to publish something to crates.io? And I didn't find any release or build workflows which is related to the `wasmedge-macro` crates.

File: bindings/rust/wasmedge-macro/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        match sys_async_expand_host_func(&item_fn) {
+            Ok(token_stream) => token_stream.into(),
+            Err(err) => err.to_compile_error().into(),
+        }
+    } else {
+        TokenStream::new()
+    }
+}
+
+fn sys_async_expand_host_func(item_fn: &syn::ItemFn) -> syn::Result<proc_macro2::TokenStream> {
```
Comment (hydai) [inline bindings/rust/wasmedge-macro/src/lib.rs:165]: What's the naming rule?
I found `expand_async_host_func2` and `async_host_func2`, however, these are `sys_async_host_function` and `sys_async_host_function`.

**Review by hydai — CHANGES_REQUESTED** [review]

**Comment by apepkuss** [comment]
> Okay, so this is the expected behavior of the current workflows. Just another thing to check, once this PR got merged, do I need to publish something to crates.io? And I didn't find any release or build workflows which is related to the `wasmedge-macro` crates.

For now, I don't think we need to publish `wasmedge-macro`, as it's still under development. In addition, this PR only uses the `host_function` macro, while others are defined for async cases that could be launched in future PRs. According to the plan, the `wasmedge-macro` crate will be published as part of the next Rust bindings release.

**Comment by hydai** [comment]
> For now, I don't think we need to publish `wasmedge-macro`, as it's still under development. In addition, this PR only uses the `host_function` macro, while others are defined for async cases that could be launched in future PRs. According to the plan, the `wasmedge-macro` crate will be published as part of the next Rust bindings release.

Thanks for clarifying this.

File: bindings/rust/wasmedge-macro/src/lib.rs (hunk 1 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-macro/src/lib.rs:165]: The `async_host_func2` macro is defined for declaring async host function when using `sdk`, while `sys_async_host_function` and `sys_async_host_function2` are two different macro types and used for declaring async host function when using `sys`. These three macros are used in another branch. As for `expand_async_host_func2`, it is just a private function used by the `async_host_func2` macro, and not exposed to the outside. Thanks!

**Comment by apepkuss** [comment]
> > For now, I don't think we need to publish `wasmedge-macro`, as it's still under development. In addition, this PR only uses the `host_function` macro, while others are defined for async cases that could be launched in future PRs. According to the plan, the `wasmedge-macro` crate will be published as part of the next Rust bindings release.
> 
> Thanks for clarifying this.

My pleasure :)

**Review by hydai — APPROVED** [review]

**Comment by apepkuss** [comment]
@hydai Thanks for the review!

**Pull request merged**